### PR TITLE
Introducing `AutoPeftModelForxxx`

### PR DESCRIPTION
**Pull request by younesbelkada**
This PR introduces a new paradigm, `AutoPeftModelForxxx` intended for users that want to rapidly load and run peft models.
Currently a user needs to run all these steps:
```python
from peft import PeftConfig, PeftModel
from transformers import AutoModelForCausalLM

peft_config = PeftConfig.from_pretrained("ybelkada/opt-350m-lora") 
base_model_path = peft_config.base_model_name_or_path

transformers_model = AutoModelForCausalLM.from_pretrained(base_model_path, device_map="auto", load_in_8bit=True)
peft_model = PeftModel.from_pretrained(transformers_model, peft_config)
```
to load a peft model from the Hub or locally, whereas now they could just do:
```python
from peft import AutoPeftModelForCausalLM

peft_model = AutoPeftModelForCausalLM.from_pretrained("ybelkada/opt-350m-lora")
```

cc @pacman100 @BenjaminBossan 

TODOs:
- [x] add tests
- [x] add docs

### CODE REVIEW

File: tests/test_auto.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    PeftModelForCausalLM,
+    PeftModelForFeatureExtraction,
+    PeftModelForQuestionAnswering,
+    PeftModelForSeq2SeqLM,
+    PeftModelForSequenceClassification,
+    PeftModelForTokenClassification,
+)
+
+
+class PeftAutoModelTester(unittest.TestCase):
```
Comment (BenjaminBossan) [inline tests/test_auto.py:36]: The individual tests all look very similar. I wonder if they could be parametrized.

**Review by BenjaminBossan — APPROVED** [review]
This looks quite good to me, I have only minor comments.

Admittedly, I'm not an expert in the concept of the Auto* models, so I can't comment on the overall design of this feature. Let's see what Sourab has to add to this.

File: tests/test_auto.py (hunk 1 above)
Comment (younesbelkada) [reply tests/test_auto.py:36]: Yeah they are quite similar however, to check if the models are effectively converted in `bfloat16` for instance I check some custom module attributes for each case. Maybe let's keep that as it is

File: src/peft/auto.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+)
+from .utils import PeftConfig
+
+
+class _BaseAutoPeftModel:
+    _target_class = None
+    _target_peft_class = None
+
+    def __init__(self, *args, **kwargs):
+        raise EnvironmentError(
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/auto.py]: Done in [60eef55](https://github.com/huggingface/peft/pull/694/commits/60eef559852b815c9ccce2108914b9fe5933159b)

File: src/peft/auto.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+)
+from .utils import PeftConfig
+
+
+class _BaseAutoPeftModel:
+    _target_class = None
+    _target_peft_class = None
+
+    def __init__(self, *args, **kwargs):
+        raise TypeError(
```
Comment (BenjaminBossan) [inline src/peft/auto.py:46]: Oh, I see now that transformers uses `EnvironmentError` and that's why you used it. I still think it's the wrong error to raise, but _maybe_ it should be used for consistency _if_ there is some code somewhere that catches this error specifically??

**Review by pacman100 — COMMENTED** [review]
Great work @younesbelkada, I like the simplified UX. I have a few queries based on the offline discussion.

1. Methods such as LoRA, AdaLora, IA3, AdaptionPrompt etc are widely applicable to tasks other than those explicitly supported as they do inline changes to the modules, such as ASR, Image Captioning using BLIP, Stable DIffusion, etc. What are your thoughts on going about it as users might expect it to work for those too? Or do we restrict this API to only the explicitly supported NLP tasks?
2. Left a comment

File: src/peft/auto.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+            f"{self.__class__.__name__} is designed to be instantiated "
+            f"using the `{self.__class__.__name__}.from_pretrained(pretrained_model_name_or_path)` or "
+            f"`{self.__class__.__name__}.from_config(config)` methods."
+        )
+
+    @classmethod
+    def from_pretrained(cls, pretrained_model_name_or_path, *peft_model_args, **kwargs):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/peft/auto.py:53]: Thanks for the valuable feedback, yes that would makes totally sense

File: src/peft/auto.py (hunk 3 above)
Comment (younesbelkada) [reply src/peft/auto.py:46]: I see that makes sense, will revert it to `EnvironmentError` for consistency with transformers!

**Review by younesbelkada — COMMENTED** [review]
Hi @pacman100 
Thanks for your review, regarding your first point, we could probably do a first version with only officially supported NLP tasks and do a second iteration to add a new auto mapping class that group the model classes based on modalities

File: src/peft/auto.py (hunk 3 above)
Comment (BenjaminBossan) [reply src/peft/auto.py]: Maybe add a comment that explains why.

Comment (younesbelkada) [reply src/peft/auto.py]: Sure, just added a comment

**Review by pacman100 — APPROVED** [review]
Thank you @younesbelkada for iterating, LGTM!

**Pull request merged**